lib/sqlalchemy/dialects/mysql/base.py

```diff
@@ -1494,9 +1495,28 @@ def visit_join(self, join, asfrom=False, from_linter=None, **kwargs):
 
     def for_update_clause(self, select, **kw):
         if select._for_update_arg.read:
-            return " LOCK IN SHARE MODE"
+            tmp = " LOCK IN SHARE MODE"
         else:
-            return " FOR UPDATE"
+            tmp = " FOR UPDATE"
+
+        if select._for_update_arg.of and self.dialect._is_mysql:
+
+            tables = util.OrderedSet()
+            for c in select._for_update_arg.of:
```

Comment:
**Gord Thompson** (gordthompson) wrote:

This test fails because the `server_version_info` is None. Debugging shows that on first(?) connect `__get_server_version_info()` is called and the return value is saved to `server_version_info` for a particular dialect object (e.g., <sqlalchemy.dialects.mysql.mysqldb.MySQLDialect_mysqldb object at 0x0000007BD0A7FD48>). However, when this code runs, `self.dialect` is a *different* dialect object (e.g., <sqlalchemy.dialects.mysql.mysqldb.MySQLDialect_mysqldb object at 0x0000007BD2127308>) which has clobbered the `server_version_info` value.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1928

Reply:
**Federico Caselli** (CaselIT) wrote:

I think you need to set the sets as __backend__=True and connect fist at least one, so that the dialect initialization happens.
I don't know it there is a way of mocking the server version in the test.
It think it would be useful

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1928

Reply:
I meand tests, not sets

Reply:
**Gord Thompson** (gordthompson) wrote:

Sorry for the confusion caused by my incorrect wording. Some tests are failing, but they are failing because of this code in the mysql dialect itself. And like I said, the PyCharm debugger shows that the correct server version values are being retrieved (`SELECT VERSION()`) and inserted into some dialect object, but not the same dialect object that is being referenced here.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1928

Reply:
**mike bayer** (zzzeek) wrote:

lets give the MySQL dialect a flag:

class MySQLDialect(...)
   # ...
   supports_for_update_of = False


compiler only checks self.dialect.supports_for_update_of.

MySQL dialect when it does on_connect (sorry forgot event name), has server version, can update this flag.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1928

Reply:
**Gord Thompson** (gordthompson) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1928